### ISSUE TITLE
Make `encoded_len` const

### ISSUE THREAD

**Issue by edgarogh** [marshallpierce/rust-base64#251]
# The problem

- In `no-std` contexts where allocation is impossible or best avoided, it can be useful to allocate an array on the stack whose size is the maximum possible required size to encode a value. For instance, if a value can be at most `256` bytes long, it may be desirable to stack-allocate a `[0u8; 344]` upfront. bviously, that's only possible if the value is small enough and/or the call stack is large enough.
- Hardcoding a magic number like this "344" hinders readability, and `base64::` already has an `encoded_len` function that does the job perfectly, albeit not in `const` contexts.

# How I, the issue filer, am going to help solve it

Here's a minimal implementation (note that `unreachable!` can't contain a format string in const contexts):
```diff rust
-pub fn encoded_len(bytes_len: usize, padding: bool) -> Option<usize> {
+pub const fn encoded_len(bytes_len: usize, padding: bool) -> Option<usize> {
     let rem = bytes_len % 3;

     let complete_input_chunks = bytes_len / 3;
-    let complete_chunk_output = complete_input_chunks.checked_mul(4);
+    let Some(complete_chunk_output) = complete_input_chunks.checked_mul(4) else {
+        return None;
+    };

     if rem > 0 {
         if padding {
-            complete_chunk_output.and_then(|c| c.checked_add(4))
+            complete_chunk_output.checked_add(4)
         } else {
             let encoded_rem = match rem {
                 1 => 2,
                 2 => 3,
-                _ => unreachable!("Impossible remainder"),
+                _ => unreachable!(), // Impossible remainder
             };
-            complete_chunk_output.and_then(|c| c.checked_add(encoded_rem))
+            complete_chunk_output.checked_add(encoded_rem)
         }
     } else {
-        complete_chunk_output
+        Some(complete_chunk_output)
     }
 }
```

And [a brute-force test from 0 to 1000](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=5ca0d437dfb80ede4f52c7a0de40bd1b).

# Other approaches

If an `encoded_len_const` function is defined but not exported as-is, we may also directly want a
```rust
type LargeEnoughEncodingBuffer<const DATA_MAX_LEN> = [u8; encoded_len_const(DATA_MAX_LENGTH)];
```
...while we're at it (name should be changed :sweat_smile:).

**Comment by marshallpierce** [comment]
What a good idea, and I happen to have time at the moment to do something about it. I'll release this shortly.

**Comment by marshallpierce** [comment]
Released in 0.21.4

**Comment by edgarogh** [comment]
That's amazing, thank you so much!
